src/utils/allocator.h

```diff
@@ -5,11 +5,30 @@ extern const char *_text_malloc_error;
 extern const char *_text_calloc_error;
 extern const char *_text_realloc_error;
 
+// Allocator options
+#define ALLOC_HINT_TEXTURE 0x00000001
+
 // Add ifdefs here to include platform-specific allocators.
+#ifdef N64_BUILD
+#include "utils/allocator_n64.h"
+#else
 #include "utils/allocator_default.h"
+#endif
 
 #define omf_malloc(size) omf_malloc_real((size), __FILE__, __LINE__)
 #define omf_calloc(nmemb, size) omf_calloc_real((nmemb), (size), __FILE__, __LINE__)
 #define omf_realloc(ptr, size) omf_realloc_real((ptr), (size), __FILE__, __LINE__)
 
+// omf_alloc_with_options, this function allocates memory that is to be shared between the engine (on the CPU) and the
+// hardware directly. These allocations can be textures, palettes, audio buffers or pointers to read only memory.
+// The options flags are hints as to what type of allocation is necessary, it is up to the platform's implementation to
+// decide how to translate the flags: what alignment is necessary to be optimal and whether memory needs to be allocated
+// cached or uncached.
+// Using this function can allow the platform to save on copies and translations in the platform hardware implementation
+// layers.
+//
+// NOTE: An allocation with a specific set of options also requires a matching free with the same set of options.
+//
+#define omf_alloc_with_options(nmemb, size, options)                                                                   \
```

Comment:
If the options is going to declare different types of allocations, then what happens if ALLOC_HINT_TEXTURE and e.g. ALLOC_HINT_SOUND are defined simultaneously ? This is a very dangerous thing for undefined behavirour. Perhaps the flags should be an enum instead.

Reply:
It's probably worth thinking about what allocator flags we'd actually want to support. Obviously some combinations are nonsensical (like TEXTURE | SOUND) but we could also add a simple check, in debug mode, for invalid flag combinations, if we wished.

Reply:
Srsly, it might be best to split this to functions by purpose, and then allow more detailed flags for those as required. So malloc_texture(amount, flags), malloc_sound(amount, flags) etc. That would prevent nonsensical combinations, and just reserve the options flag for fine-grained changes.

Reply:
the other use, other than n64 support, that I can think of would be an arena style allocator, where we could indicate whether an allocation should live for the life of the scene (in the scene arena allocation), in a global/engine arena or should be manually managed (a more ephemeral object). Then we could use the hints to track the kind of allocation we're doing.

Reply:
I'd really like to use some arena allocator also. If someone has the interest, I would be all for integrating one.